extensions/amp-stream-gallery/0.1/test/test-amp-stream-gallery.js

```diff
@@ -0,0 +1,706 @@
+}
+
+// Use an empty gif as the image source, since we do not care what it is.
+const IMG_URL =
+  'data:image/gif;base64,' +
+  'R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
+
+describes.realWin(
+  'amp-stream-gallery',
+  {
+    amp: {
+      extensions: ['amp-stream-gallery'],
+    },
+  },
+  env => {
+    let win;
+    let doc;
+    let container;
+
+    beforeEach(() => {
+      win = env.win;
+      doc = win.document;
+      env.iframe.width = '1000';
+      env.iframe.height = '1000';
+      container = doc.createElement('div');
+      doc.body.appendChild(container);
+
+      toggleExperiment(win, 'amp-stream-gallery', true);
+    });
+
+    afterEach(() => {
+      doc.body.removeChild(container);
+    });
+
+    /**
+     * @param {!Element} el
+     * @return {!Array<!Element>}
+     */
+    function getItems(el) {
+      return toArray(el.querySelectorAll('[role="list"] > [role="listitem"]'));
+    }
+
+    /**
+     * @param {!Element} el
+     * @return {!Element}
+     */
+    function getScrollContainer(el) {
+      return el.querySelector('.i-amphtml-carousel-scroll');
+    }
+
+    async function getGallery({
+      slideCount = 5,
+      customArrows = false,
+      width,
+      attrs = {},
+    } = {}) {
+      const el = doc.createElement('amp-stream-gallery');
+      el.setAttribute('layout', 'fixed');
+      el.setAttribute('width', width);
+      el.setAttribute('height', '200');
+
+      for (const attr in attrs) {
+        el.setAttribute(attr, attrs[attr]);
+      }
+
+      for (let i = 0; i < slideCount; i++) {
+        const img = doc.createElement('amp-img');
+        img.setAttribute('src', IMG_URL);
+        img.setAttribute('width', '200');
+        img.setAttribute('height', '200');
+        el.appendChild(img);
+      }
+
+      if (customArrows) {
+        const prevArrow = document.createElement('div');
+        prevArrow.className = 'custom-arrow';
+        prevArrow.setAttribute('slot', 'prev-arrow');
+        setStyles(prevArrow, {
+          width: '60px',
+          height: '60px',
+        });
+        el.appendChild(prevArrow);
+
+        const nextArrow = document.createElement('div');
+        nextArrow.className = 'custom-arrow';
+        nextArrow.setAttribute('slot', 'next-arrow');
+        setStyles(nextArrow, {
+          width: '60px',
+          height: '60px',
+        });
+        el.appendChild(nextArrow);
+      }
+
+      container.appendChild(el);
+
+      await el.build();
+      await el.layoutCallback();
+      await afterIndexUpdate(el);
+
+      return el;
+    }
+
+    describe('rendering', () => {
+      describe('inset arrows', () => {
+        it('should be correct for default arrows', async () => {
+          const el = await getGallery({
+            slideCount: 5,
+            width: 800,
+            attrs: {
+              'min-item-width': '200',
+            },
+          });
+
+          expect(toArray(getItems(el))).to.have.length(5);
```

Comment:
This question may be irrelevant, but I'm not certain since I have more experience with Jest on `ampproject/amp-github-apps` than with Mocha in this repo. When `async` functions fail, they can raise an error and the `expect` is never called. In Jest, this can result in tests passing without ever executing the `expect`s, so tests either include `expect.assertions(N)` or 

```
it('...', async done => {
  ... await ...
  done();
});
```

How does Mocha handle async assertions like this?

Reply:
In Mocha, if a function returns a `Promise` (which all `async` functions do), then it waits for that `Promise` to resolve, or for the timeout to be hit. If the `await` never occurs, then the test fails with a timeout.

Mocha also supports the done pattern (useful if you aren't using Promises), but it isn't really needed for `async` tests.

Reply:
Gotcha, thanks for explaining!